### PR TITLE
[1.0] Refactor config migrations and remove config parameters

### PR DESCRIPTION
**Pull request by strawgate**
Fixes: #271
Fixes: https://github.com/legrego/homeassistant-elasticsearch/issues/287
Fixes: https://github.com/legrego/homeassistant-elasticsearch/issues/273

Removes `config`, removes reading config values from anything other than init (and maybe async_init), and moves all components to explicitly referencing either data or options. Also removes the configuration that stops all tests from running after the first failure. Also fixes ES setup on Linux

### CODE REVIEW

**Comment by strawgate** [comment]
@legrego one of the things im thinking about doing in this PR is to stop using build_full_config and to start using build_data and build_options instead to make it explicit where these are coming from as options are currently duplicated between data and options.

This will probably mean I have to touch most areas in the component and a lot of tests so I want to check in with you before I do that.

**Comment by legrego** [comment]
> @legrego one of the things im thinking about doing in this PR is to stop using build_full_config and to start using build_data and build_options instead to make it explicit where these are coming from as options are currently duplicated between data and options.
> 
> 
> 
> This will probably mean I have to touch a lot of tests so I want to check in with you before I do that.

@strawgate That makes sense to me, I'd be happy for you to do this. Thanks!

**Comment by strawgate** [comment]
@legrego and maybe drop yaml config? :)

**Comment by legrego** [comment]
I'm not prepared to drop yml config just yet 🙂. I believe that's still being used in the wild

**Comment by strawgate** [comment]
> I'm not prepared to drop yml config just yet 🙂. I believe that's still being used in the wild

that should be fine though right? doesn't the yml config get imported into data/options on each run?

So the only thing that would break would be users adding new key/values to their yml config?

**Comment by strawgate** [comment]
As I'm getting started on this I'm also realizing that we could probably stop accepting a `config` param any place that the config_entry is available? Does that sound right?

**Comment by legrego** [comment]
> > I'm not prepared to drop yml config just yet 🙂. I believe that's still being used in the wild
> 
> 
> 
> that should be fine though right? doesn't the yml config get imported into data/options on each run?
> 
> 
> 
> So the only thing that would break would be users adding new key/values to their yml config?

I _think_ the component will fail to start if there is unexpected yml config in the file. That's how it behaved in the past anyway.

**Comment by legrego** [comment]
> As I'm getting started on this I'm also realizing that we could probably stop accepting a `config` param any place that the config_entry is available? Does that sound right?

Yes! That's been on my mental todo list for awhile.

**Comment by strawgate** [comment]
This commit shows my rough plan for removing `config` and explicitly pulling data and options:

https://github.com/legrego/homeassistant-elasticsearch/pull/272/commits/07508e3f843253405f750de53f3453e6a8cf6a4e

It looks like `config` was passed in a couple of places to allow things like testing the gateway without committing the settings to the config_entry.

In the case of the gateway I've added some static methods to es_gateway to allow testing connections without "mocking" a config object.

**Comment by legrego** [comment]
> This commit shows my rough plan for removing `config` and explicitly pulling data and options:
> 
> [07508e3](https://github.com/legrego/homeassistant-elasticsearch/commit/07508e3f843253405f750de53f3453e6a8cf6a4e)
> 
> It looks like `config` was passed in a couple of places to allow things like testing the gateway without committing the settings to the config_entry.
> 
> In the case of the gateway I've added some static methods to es_gateway to allow testing connections without "mocking" a config object.

The general approach LGTM. Is the plan to remove the `ESPrivilegeCheck` class in favor of the privilege checks you're embedding into the gateway?

**Comment by strawgate** [comment]
Fixes: https://github.com/legrego/homeassistant-elasticsearch/issues/287
Fixes: https://github.com/legrego/homeassistant-elasticsearch/issues/273

**Comment by strawgate** [comment]
apologies in advance...

**Review by legrego — COMMENTED** [review]
This looks fantastic, thanks for the refactor! The parameterized tests and snapshots are much easier to reason about as well 👏

File: custom_components/elasticsearch/__init__.py
Comment (legrego) [reply custom_components/elasticsearch/__init__.py:252]: Thanks for explaining your view -- I can understand where you're coming from, and I don't have a strong enough opinion to force this change. Feel free to leave this alone if that's your preference.

File: custom_components/elasticsearch/__init__.py
Comment (legrego) [reply custom_components/elasticsearch/__init__.py:267]: I'm not married to either approach. IMO introducing a config property to describe what they've configured is a proxy for the actual source of truth. This is less accurate in my mind than just seeing which values are configured.

> refactor places that used to do these checks to only have to check auth method using these CONF_AUTH constants.

It seems we haven't fully realized this goal?

https://github.com/strawgate/homeassistant-elasticsearch/blob/e13b6160addf329e0ccaaf23f59d4a58335a5e47/custom_components/elasticsearch/es_gateway.py#L283-L302

Comment (strawgate) [reply custom_components/elasticsearch/__init__.py:267]: I could refactor that too.

My main goal was to stop having empty string attached to CONF_USERNAME when the using an API key.

I could probably revert back to not having an auth_mode and keep the change that prevents populating the username/password keys at all when api keys are used and vice-versa. That would simplify the checks to be one line each (if conf_username, if conf_api_key, else no authentication)

it's just whether we want a proxy variable or to rely on a side effect of the user picking an authentication type

Comment (strawgate) [reply custom_components/elasticsearch/__init__.py:263]: Remove auth_method in latest commit

**Comment by strawgate** [comment]
Addressed / incorporated all feedback

**Review by legrego — APPROVED** [review]
LGTM once merge conflict is addressed

**Comment by strawgate** [comment]
!!!!

**Pull request merged**
